dep0/lib/03_typecheck/src/type_assign.cpp

```diff
@@ -0,0 +1,332 @@
+                            err.str(),
+                            loc,
+                            std::vector<dep0::error_t>{
+                                std::move(lhs.error()),
+                                std::move(rhs.error())
+                            }));
+                    }
+                });
+        },
+        [&] (parser::expr_t::var_t const& x) -> expected<expr_t>
+        {
+            auto const var = expr_t::var_t{x.name};
+            auto const val = ctx[var];
+            if (not val)
+            {
+                std::ostringstream err;
+                pretty_print<properties_t>(err << "unknown variable `", var) << '`';
+                return error_t::from_error(dep0::error_t(err.str(), loc));
+            }
+            return make_legal_expr(
+                match(
+                    val->value,
+                    [&] (type_def_t const&) -> sort_t { return derivation_rules::make_typename(); },
+                    [&] (expr_t const& expr) -> sort_t { return expr.properties.sort.get(); }),
+                var);
+        },
+        [&] (parser::expr_t::app_t const& x) -> expected<expr_t>
+        {
+            return type_assign_app(ctx, x, loc);
+        },
+        [&] (parser::expr_t::abs_t const& f) -> expected<expr_t>
+        {
+            return type_assign_abs(ctx, f, loc, std::nullopt);
+        },
+        [&] (parser::expr_t::pi_t const& pi) -> expected<expr_t>
+        {
+            auto pi_ctx = ctx.extend();
+            return check_pi_type(pi_ctx, pi.args, pi.ret_type.get()); // TODO should this be renamed too?
+        },
+        [&] (parser::expr_t::array_t const&) -> expected<expr_t>
+        {
+            // the elemnt type of the array must be of sort types,
+            // i.e. a "function argument" of type `typename`;
+            // to have array of types (eg `{int, bool, bool}`),
+            // we fisrt need to add cumulativity of types;
+            // i.e. we can have an array of ints, eg {1,2,3},
+            // but not an array of types, eg {int, bool, bool};
+            // with cumulativity we can have both
+            return make_legal_expr(
+                make_legal_expr(
+                    kind_t{}, // TODO need to add a test to make sure this is correct
+                    expr_t::pi_t{
+                        std::vector{
+                            make_legal_func_arg(derivation_rules::make_typename()),
+                            make_legal_func_arg(derivation_rules::make_u64())
+                        },
+                        derivation_rules::make_typename()}),
+                expr_t::array_t{});
+        },
+        [&] (parser::expr_t::init_list_t const& init_list) -> expected<expr_t>
+        {
+            std::ostringstream err;
+            err << "cannot assign a unique type to an initializer list expression";
+            return error_t::from_error(dep0::error_t(err.str(), loc));
```

Comment:
Well I could use the same technique I used for arithmetic expressions, ie if I can assign a type to any value, I can then check all other against that type, which will also produce a vector of reasons

Reply:
No, that's not the reason why we can't assign a type to an initializer-list, but the fact that, in general, they can be used to construct other types besides arrays, eg units and structs, even though for now we can only construct arrays.